### PR TITLE
Revise GCM message sending and migrate to FCM

### LINKED ISSUE #182: Google Cloud Messaging deprecated, will be removed no later than April 2019
Citing the e-mail I got from Google some days ago:

> In 2016, we launched Firebase Cloud Messaging (FCM), the successor to Google Cloud Messaging (GCM). As the next evolution of GCM, Firebase Cloud Messaging allows you to send notifications and data messages reliably to iOS, Android, and the Web at no cost. In addition, FCM provides you with new features like the easy-to-use notifications interface in the Firebase console, so you can easily target and test notifications to re-engage your users.
> In order to devote more time and attention to improving FCM, today we’re announcing that you must upgrade to FCM in the next year. The GCM server and client APIs are deprecated and will be removed as soon as April 11, 2019. We recommend you upgrade sooner rather than later so you can start taking advantage of the new features in FCM today.
> If you have projects that are still using the GCM APIs, you will need to update your client and server code to use FCM before April 11, 2019. Your existing GCM tokens will continue to work with FCM so you won’t lose the ability to send to your existing users.
> What do I need to do?
> You’ll find a complete walkthrough of the FCM upgrade process here. You can also check out this video, which walks through the migration process.
> If you have questions that aren’t answered in the deprecation documentation, take a look at the Cloud Messaging FAQs on our support page or reach out to us through one of our other support channels.
> We look forward to welcoming you to FCM!
> Thanks,
> Firebase Team

The migration guide can be found [here](https://developers.google.com/cloud-messaging/android/android-migrate-fcm?utm_medium=email&utm_source=msa). We probably should take care of migrating the server implementation from GCM to FCM in order to be able to send messages after GCM is shut down.

This could also be a good opportunity to revise the whole GCM sending code stuff in this project, so it may be easier to understand.

### PR DESCRIPTION
**Pull request by FlorianSW**
Fixes #182

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -61,7 +61,7 @@ var taskEnv = process.env.TASK || 'main';
 
 // If Google Cloud Messaging is configured set it up
```
Comment (mueller-ma) [inline app.js]: `If Firebase...`

File: fcm-xmpp.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
+const system = require('./system');
+
+/**
+ * This module maintains XMPP connection to GCM to receive messages from Android
```
Comment (mueller-ma) [inline fcm-xmpp.js]: `...to FCM...`

**Comment by digitaldan** [comment]
Thanks, this is awesome.  Couple questions, did you test with your own firebase account?  How does the switchover work, if we were  to deploy this, would notifications continue to work for users as-is?  Could we have one of our servers running GCM and another running firebase during the upgrade process?   In honesty i have not looked at Firebase push messaging, but it seems to support the older GCM xmpp api?

**Comment by FlorianSW** [comment]
Regarding the GCM XMPP interface: FCM supports the same endpoints as GCM and they're "virtaully" the same, see the [migration docs](https://developers.google.com/cloud-messaging/android/android-migrate-fcm#update-server-endpoints):
> FCM supports HTTP and XMPP protocols that are virtually identical to the GCM server protocols, so you don't need to update your sending logic for the migration. 

For the keys used on the server side: These are the same keys as I used for GCM before (for my local environment), however, iirc, I created these keys in firebase directly, as GCM was already deprecated at this time. So, it might be a good idea to startup an instance locally with the same keys as of production (or QS if you've one) and check, if it is working fine with these credentials, too. Otherwise you would need to follow the [migration guide](https://developers.google.com/cloud-messaging/android/android-migrate-fcm#import-your-gcm-project-as-a-firebase-project) to import the GCM project into Firebase before upgrading openhab-cloud. However, looking at how the guide is written, I would assume that this is for the client-side migration, and as I said, my credentials I used for GCM works just fine for FCM, that's why I would test it locally before :)

I also don't see, why a staged deployment on one server only wouldn't work, as GCM and FCM should be interchangeable (see my words above).

**Comment by digitaldan** [comment]
>I also don't see, why a staged deployment on one server only wouldn't work, as GCM and FCM should be interchangeable (see my words above).

Awesome, I have to travel this week, so I may not get to this or your other PR until Thursday or Friday.

**Comment by digitaldan** [comment]
Just tested on my local instance, all looks good, ready to merge on your side?

**Comment by FlorianSW** [comment]
Yep, ready to merge :)

**Comment by digitaldan** [comment]
[Cloud PR's](https://trello.com/c/Aoq5a7w1/20-cloud-prs)

**Comment by maniac103** [comment]
@digitaldan Any updates here? This will conflict heavily with #175 so either me or @FlorianSW will need to update the respective pull request after the other one is merged.

File: notificationsender/firebase.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        }
+
+        const data = {
+            type: 'notification',
+            notificationId: androidNotificationId,
+            message: message,
+        };
+        firebaseClient.message(registrationIds, data, firebaseOptions, function (result) {
+            if (result.failure) {
+                logger.error("openHAB-cloud: GCM send error: " + err);
```
Comment (digitaldan) [inline notificationsender/firebase.js]: "err" is not defined anywhere

**Comment by digitaldan** [comment]
I'm ready to merge this minus one small comment above, @maniac103 I think we will need to refactor yours in due to the timing issue with GCM.  I will shoot for deployment on Saturday/Sunday if all goes well.

**Comment by RyanRamchandar** [comment]
@digitaldan @FlorianSW any plans to cut a release so we have time to migrate before the Apr 11 deadline?

**Comment by FlorianSW** [comment]
@digitaldan is deploying and releasing it :)

**Comment by digitaldan** [comment]
This is now released, thanks!

**Comment by RyanRamchandar** [comment]
Thank you!

**Pull request merged**